### ISSUE TITLE
Onedrive invalidRequest: nameContainsInvalidCharacters

### ISSUE THREAD

**Issue by gungun974** [Jwink3101/syncrclone#11]
Hello. I found your program on the Internet to do ***two-way syncronization with Rclone***.
Your program ***works fine with Onedrive*** in the sense that the folders and files that need to be syncronized are syncronized.
***However at the end, the program crashes*** since for some reason I don't know. ***Onedrive does not support the "_" or the "-"*** (I don't remember which of the two it is ...)
***So syncrclone can't copy his internals files and crash***.
My question is is it possible to change `.syncrclone/{AB}-{name} _fl.json.xz` by something ***other than "-" or "_" without having to modify the entire file?***

***Here my logs for help :***
```
021-09-28 11:07:20: syncrclone (20210924.0.BETA)
2021-09-28 11:07:20: config path: '/home/tonecue/.syncrclone/config.py'
2021-09-28 11:07:20: A: '/home/tonecue/sync'
2021-09-28 11:07:20: B: 'onedrive:sync/'
2021-09-28 11:07:20: rclone version:
2021-09-28 11:07:20: rclone: rclone v1.57.0-DEV
2021-09-28 11:07:20: rclone: - os/version: gentoo 2.7 (64 bit)
2021-09-28 11:07:20: rclone: - os/kernel: 5.14.6-gentoo (x86_64)
2021-09-28 11:07:20: rclone: - os/type: linux
2021-09-28 11:07:20: rclone: - os/arch: amd64
2021-09-28 11:07:20: rclone: - go/version: go1.17.1
2021-09-28 11:07:20: rclone: - go/linking: dynamic
2021-09-28 11:07:20: rclone: - go/tags: none
2021-09-28 11:07:20:
2021-09-28 11:07:20: Refreshing file list on A '/home/tonecue/sync'
2021-09-28 11:07:20: 1 files, 0.00 b
2021-09-28 11:07:20: Refreshing file list on B 'onedrive:sync/'
2021-09-28 11:07:24: 1 files, 0.00 b
2021-09-28 11:07:24:
2021-09-28 11:07:24:
2021-09-28 11:07:24: Actions queued on A:
2021-09-28 11:07:24:
2021-09-28 11:07:24: Actions queued on B:
2021-09-28 11:07:24:
2021-09-28 11:07:24: Performing Actions on A
2021-09-28 11:07:24:
2021-09-28 11:07:24: Performing Actions on B
2021-09-28 11:07:26:
2021-09-28 11:07:26: A >>> B 0 files, 0.00 b
2021-09-28 11:07:26:
2021-09-28 11:07:26: A <<< B 0 files, 0.00 b
2021-09-28 11:07:26:
2021-09-28 11:07:26: No need to refresh file list on A. Updating current state
2021-09-28 11:07:26: No need to refresh file list on B. Updating current state
2021-09-28 11:07:48:  rclone stderr: 2021/09/28 11:07:39 ERROR : Attempt 1/3 failed with 1 errors and: invalidRequest: nameContainsInvalidCharacters: Path (//.syncrclone/B-t8RwY_fl.json.xz) contains invalid leading character.
2021/09/28 11:07:43 ERROR : Attempt 2/3 failed with 1 errors and: invalidRequest: nameContainsInvalidCharacters: Path (//.syncrclone/B-t8RwY_fl.json.xz) contains invalid leading character.
2021/09/28 11:07:48 ERROR : Attempt 3/3 failed with 1 errors and: invalidRequest: nameContainsInvalidCharacters: Path (//.syncrclone/B-t8RwY_fl.json.xz) contains invalid leading character.
2021/09/28 11:07:48 Failed to copyto: invalidRequest: nameContainsInvalidCharacters: Path (//.syncrclone/B-t8RwY_fl.json.xz) contains invalid leading character.

2021-09-28 11:07:48: RCLONE ERROR
2021-09-28 11:07:48: CMD ['rclone', '--onedrive-list-chunk', '5000', '--checkers', '15', 'copyto', '/tmp/tmpkrzik5ew/B_curr', 'onedrive:sync//.syncrclone/B-t8RwY_fl.json.xz']
2021-09-28 11:07:48: STDOUT
2021-09-28 11:07:48: STDERR 2021/09/28 11:07:39 ERROR : Attempt 1/3 failed with 1 errors and: invalidRequest: nameContainsInvalidCharacters: Path (//.syncrclone/B-t8RwY_fl.json.xz) contains invalid leading character.
2021/09/28 11:07:43 ERROR : Attempt 2/3 failed with 1 errors and: invalidRequest: nameContainsInvalidCharacters: Path (//.syncrclone/B-t8RwY_fl.json.xz) contains invalid leading character.
2021/09/28 11:07:48 ERROR : Attempt 3/3 failed with 1 errors and: invalidRequest: nameContainsInvalidCharacters: Path (//.syncrclone/B-t8RwY_fl.json.xz) contains invalid leading character.
2021/09/28 11:07:48 Failed to copyto: invalidRequest: nameContainsInvalidCharacters: Path (//.syncrclone/B-t8RwY_fl.json.xz) contains invalid leading character.
ERROR. Dumping logs (with debug) to '/tmp/tmp5pi6sjjv/log.txt'
2021-09-28 11:07:48: ERROR: Command '['rclone', '--onedrive-list-chunk', '5000', '--checkers', '15', 'copyto', '/tmp/tmpkrzik5ew/B_curr', 'onedrive:sync//.syncrclone/B-t8RwY_fl.json.xz']' returned non-zero exit status 1.
```
I'm not really sure how to write an issue on github so sorry if it's very messy.

**Comment by Jwink3101** [comment]
I use my tool with OneDrive _extensively_! It's not an issue with the dash or underscore. Both are supported and/or encoded by rclone. (And extremely common file name formats). 

I suspect it is something to do with the `//` I am seeing with the paths. They shouldn't be there and I'll look into that. In the meantime, can you post the `remoteA = ...` and `remoteB = ...` part of your config?

**Comment by Jwink3101** [comment]
Reading your comment again, I do still think it is the `//` especially with :

    Path (//.syncrclone/B-t8RwY_fl.json.xz) contains invalid leading character.

There *is* a bug in syncrclone that makes it do the double `/` which I fixed in `20210928.0.BETA`. But even before fixing that, I cannot replicate it with 1.56.1. I wonder if something changed in the fact that you're using 1.57.DEV (I don't know which one *specifically* so I can't test).

Please update to the latest syncrclone and see if that fixes the issue? 

If not, you'll notice when syncrclone fails it prints something like:

    ERROR. Dumping logs (with debug) to '/tmp/tmp5pi6sjjv/log.txt'

Can you please post that log?

**Comment by gungun974** [comment]
Good news. The copy works perfectly.
I'm sorry for forgetting you given the configuration if not anyone, maybe I had badly written the hosts and that generated this extra "/"?
In short here here is my configuration
```python3
"""
syncrclone

Config File

This configuration file is read as Python so things can be customized as
desired. With few exception, any missing items will go to the defaults
already specified.

Flags should always be a list.
Example: `--exclude myfile` will be ['--exclude','myfile']

This is *ALWAYS* evaluated from the parent of this file.

"""
## Remotes:

# Specify the remotes to be used in the rclone config. This should be the
# root to be synchronized. If local, no need to specify a remote.
# For example:
#
#    remoteA = "/full/path/to/local"
#    remoteB = "b2:bucket/"
#
#  See docs/config_tips.md for some helpful tips and tricks
remoteA = "/home/tonecue/sync"
remoteB = "onedrive:sync/"

# Names are needed so that one directory can sync to OR from multiple
# locations. Names should be unique. The default below is randomly
# set.
name = 't8RwY'

## rclone flags

# Specify the path to the rclone executable.
rclone_exe = 'rclone'

# Specify FILTERING flags only. Note that if filtering flags are used later,
# it *will* cause issues. Examples of rclone filters:
#     --include --exclude
#     --include-from --exclude-from
#     --filter --filter-from
#
# note that ['--filter','- .syncrclone/**'] is automatically prepended.
#
# These should be specified as a list. For example, to exclude *.exc, do
# ['--exclude','*.exc']
#
# See warnings in the readme about --exclude-if-present
filter_flags = []

# General rclone flags are called every time rclone is called. This is how
# you can specify things like the conifg file.
# Remember that this config is evaluated from its parent directory.
#
# Example: ['--config','path/to/config']
#
# Note: Not all flags are compatible and may break the behavior of syncrclone
#       such as ones changing the display (`--progress`, `--log-level`)
#
#       Also, some flags may be needed for certain remotes. See
#           https://github.com/Jwink3101/syncrclone/issues/1
#       for a discussion of needing `--drive-skip-gdocs` for gdrive
#
# There is extemely minimal validation of flags. If you're uncertain, have
# a backup and test with `--dry-run`
rclone_flags = ['--onedrive-list-chunk' , '5000' , '--checkers' , '15']

# The following are added to the existing environment.
# These should NOT include any filtering!
rclone_env = {'RCLONE_CONFIG': '/home/tonecue/.config/rclone/rclone.conf'}

# Additionally, specify flags for only one side or the other.
# Examples would be things like `--fast-list`
rclone_flagsA = []
rclone_flagsB = []

## Sync Options

# How to compare files on A and B. Note that mtime also includes size.
# Hashes will be automatic based on common hashes. Will raise an error if
# there are no common hashes.
# Options: {'size','mtime','hash'}
#
# Notes:
#   * if using size, can have false negatives.
compare = 'mtime'

# When doing mtime comparisons, what is the error to allow
dt = 1.1 # seconds

# How to handle conflicts.
# Note that even if comparison is done via hash, you can still resolve via
# mod time. Be aware that not all remotes return reliable mod times and adjust
# accordingly. See https://rclone.org/overview/
#
#   'A','B'             : Always select A or B
#   'tag'               : Tag both. Makes tag_conflict irrelevant
#
#   'older','newer'     : Select the respective file. See note below.
#   'smaller','larger'  : Select the smaller or larger file
#
# If a conflict cannot be resolved it will default to 'tag' and print a warning.
conflict_mode = 'newer'

# You can choose to tag the other file rather than overwrite it. If tagged,
# it will get renamed to have appended `.{time}.{A or B}` to the file
tag_conflict = False

# Hashes can be expensive to compute on some remotes such as local or sftp.
# As such, rather than recompute them all, the hashes of the previous state
# can be used if the filename,size,mtime all match. Then the hashes of the
# remaining files will be computed in a second rclone call.
#
# For most other remotes (e.g. S3, B2), hashes are stored by the remote
# so there is no need to reuse them
reuse_hashesA = False
reuse_hashesB = False

# Some remotes (e.g. S3) require an additional API call to get modtimes. If you
# are comparing with 'size' of 'hash', you can forgo this API call by setting
# this to False. Future versions may be smart about this and allow for
# server-side modtime with a cache but that is not yet possible.
always_get_mtime = True

# When backups are set, all overwrites or deletes will instead be backed up (moved into
# the .syncrclone folder)
backup = True

# Specify whether to also sync the backed up files between A and B. If True,
# both remotes will have all of the backups. If False, each remote will only
# have what it backed up on the respective side
sync_backups = False

# Some remotes do not support hashes at all (e.g crypt) while others do not
# always return a hash for all files (e.g. S3). When this is encountered,
# syncrclone can fall back to another `compare` or `renames{AB}` attribute.
# Specify as None (default) to instead throw an error.
hash_fail_fallback = None # {'size','mtime',None}

# By default, syncrclone will set a lock to prevent other sync jobs from
# working at the same time. Note that other tools will not respect these
# locks. They are only for syncrclone.
#
# They are not strictly needed and require extra rclone calls. As such, they
# can be disabled. Note even if setting locks if disabled, syncrclone will
# still respect them!
set_lock = False


# While transfers will follow the respective rclone flags  (e.g. ['--transfers','10'],
# delete, backup, and move actions need more calls. There is some optimization but it
# still may need more than one call. This allows it to happen in separate rclone calls.
action_threads = 1 # Some remotes do not like concurrent rclone calls so this is the default
# action_threads = __CPU_COUNT__ // 1.5
# action_threads = 4

# syncrclone does not transfer empty directories however if a directory is
# empty after a sync and it was NOT empty before (e.g. the directory was moved
# or deleted), then it can remove them. Note that (a) this only removes
# directories that were *made* empty by syncrclone, and (b) if files still
# exist in the directory (e.g. they were excluded), it will *not* delete them.
#
# This settings doesn't make sense for some remotes. Leave as None to set
# automatically based on whether the remote supports empty directories.
cleanup_empty_dirsA = None
cleanup_empty_dirsB = None

## Rename Tracking

# Renames can be tracked if the file is unmodified on both sides and only
# renamed on one side (if it is renamed to the same on both sides, it won't
# do anything). Will not track if a rename cannot be uniquely identified
#
# Tracking is done via the following. Note that the pool of considered
# files are *only* those that have been identified as new.
#
#   'size'    : Size of the file only. VERY UNSAFE
#   'mtime'   : mtime and size. Slightly safer than size but still risky
#   'hash'    : Hash of the files
#   'inode'   : inode (plus mtime and size) of the file. This is only supported for local
#               remotes and may have issues with links. Support may be removed in the
#               future.
#    None     : Disable rename tracking
#
# Because moving files has to be done with individual rclone calls, it is often more
# efficient to disable rename tracking as a delete and copy can be more efficient for
# lots of files. It also doesn't make sense to use renames on A or B if the remote B or A
# doesn't support server-side copy or move.
renamesA = None
renamesB = None

## Logs
# All output is printed to stdout and stderr but this can also be saved and
# uploaded to a remote. Note that the last upload step will not be in the logs
# themselves. The log name is fixed as '{name}_{date}.log' but the path
# can be set. Note that the `.syncrclone` folder is always excluded if you do not
# wish to also track the logs, though it can be helpful. Especially if there
# is more than one syncpair
#
# Can also specify a local location. Useful if both remotes are remote. Recall
# that the paths are relative to this file. If blank, will not save logs
log_dest = '.syncrclone/logs' # Relative to root of each remote
local_log_dest = '' # Note on a remote

## Pre- and Post-run
# Specify shell code to be evaluated before and/or after running syncrclone. Note
# these are all run from the directory of this config (as with everything else).
# STDOUT and STDERR will be captured. Note that there is no validation or
# security of the inputs. These are not actually called if using dry-run.
pre_sync_shell = ""
post_sync_shell = ""




#######
# This should only be changed by the user when migrating from an older config
# to a newer one. Just because the current version of syncrclone and the version
# below do not match, it does not mean the sync won't work.
_syncrclone_version = '20210924.0.BETA'
```

But also so here are the logs of the copy that works well under the latest version.

```
2021-09-28 20:42:43: syncrclone (20210928.0.BETA)
2021-09-28 20:42:43: config path: '/home/tonecue/.syncrclone/config.py'
2021-09-28 20:42:43: A: '/home/tonecue/sync'
2021-09-28 20:42:43: B: 'onedrive:sync/'
2021-09-28 20:42:43: rclone version:
2021-09-28 20:42:43: rclone: rclone v1.57.0-DEV
2021-09-28 20:42:43: rclone: - os/version: gentoo 2.7 (64 bit)
2021-09-28 20:42:43: rclone: - os/kernel: 5.14.6-gentoo (x86_64)
2021-09-28 20:42:43: rclone: - os/type: linux
2021-09-28 20:42:43: rclone: - os/arch: amd64
2021-09-28 20:42:43: rclone: - go/version: go1.17.1
2021-09-28 20:42:43: rclone: - go/linking: dynamic
2021-09-28 20:42:43: rclone: - go/tags: none
2021-09-28 20:42:43:
2021-09-28 20:42:43: Refreshing file list on A '/home/tonecue/sync'
2021-09-28 20:42:43: 1 files, 0.00 b
2021-09-28 20:42:43: Refreshing file list on B 'onedrive:sync/'
2021-09-28 20:42:48: 1 files, 0.00 b
2021-09-28 20:42:48:
2021-09-28 20:42:48:
2021-09-28 20:42:48: Actions queued on A:
2021-09-28 20:42:48:
2021-09-28 20:42:48: Actions queued on B:
2021-09-28 20:42:48:
2021-09-28 20:42:48: Performing Actions on A
2021-09-28 20:42:49:
2021-09-28 20:42:49: Performing Actions on B
2021-09-28 20:42:51:
2021-09-28 20:42:51: A >>> B 0 files, 0.00 b
2021-09-28 20:42:51:
2021-09-28 20:42:51: A <<< B 0 files, 0.00 b
2021-09-28 20:42:51:
2021-09-28 20:42:51: No need to refresh file list on A. Updating current state
2021-09-28 20:42:51: No need to refresh file list on B. Updating current state
2021-09-28 20:43:10: Logs will be saved on remotes to '.syncrclone/logs/t8RwY_2021-09-28T204243.log'
2021-09-28 20:43:10: ---- END OF LOG ----
2021-09-28 20:43:10: rclone: INFO  : log: Copied (new) to: t8RwY_2021-09-28T204243.log
2021-09-28 20:43:10: rclone: INFO  :     1.592 KiB / 1.592 KiB, 100%, 0 B/s, ETA -
2021-09-28 20:43:25: rclone: INFO  : log: Copied (new) to: t8RwY_2021-09-28T204243.log
2021-09-28 20:43:25: rclone: INFO  :     1.592 KiB / 1.592 KiB, 100%, 0 B/s, ETA -
```

Lately before you thank you. At the base I used the 1.53.DEV version compiled on Gentoo, if I went on 1.57.DEV it's because of another issue that talking to optimize OneDrive.

In short thanks again for your work as fast. I did not think I thought my worries will be settled today.
If you still had questions about this incident be free to ask them.
Otherwise you could close this issue very well.

**Comment by Jwink3101** [comment]
> The copy works perfectly.

Great! Glad to hear that fixed it. I certainly can't *promise* to look at issues right away but I use this tool too so I want things fixed. And I had some bandwidth today.

> maybe I had badly written the hosts 

Nah, this was on me. There was a bug with trailing `/` but it should have been supported. Most of the tests are local and it didn't matter there. Though I also couldn't get it to fail with OneDrive but it's hardly worth worrying now

BTW, unless you have it set somewhere else, you shouldn't need `rclone_env = {'RCLONE_CONFIG': '/home/tonecue/.config/rclone/rclone.conf'}`. That is the default. Doesn't hurt though.

Good luck. Please post if there are any issues!
